HETZNER: implement the provider for Hetzner DNS Console

das7pad opened the pull request:
This PR is implementing the first revision of the Provider for the Hetzner DNS Console service.

https://www.hetzner.com/dns-console
API schema: https://dns.hetzner.com/api-docs
Supported/Non-supported records are listed here https://docs.hetzner.com/dns-console/dns/general/supported-dns-record-types (they are also listed in the API schema https://dns.hetzner.com/api-docs#operation/CreateRecord)

It was pretty straight forward to implement (for a non-gopher), thanks for the detailed instructions.

Their API supports bulk creation and updating, but not bulk deletion. A follow-up PR could group corrections, but I am afraid this will add a lot of complexity, considering that we probably have to chunk them too.

Some record types are currently marked as not supported, incl. SOA records for which I added a special case in order to run the integration tests -- the clear process is not allowed to delete the existing SOA record. Most of the non-supported record types are available for updating in uploading a BIND zone file. Do you think we should change the correction mechanism of this provider to just use the BIND upload instead of incremental changes? As far as I can tell this does not play nice with records that are ignored/managed externally to dnscontrol.

In regards to maintaining this provider, I am not sure how much I can contribute in maintaining it, given that I am new to go and have very litte experience with the tooling around it. What do you think about asking the maintainers of the Hetzner go-libraries to step in? https://github.com/hetznercloud

Closes https://github.com/StackExchange/dnscontrol/issues/715

klausenbusk commented:
Are you aware of https://github.com/timohirt/terraform-provider-hetznerdns/issues/7? Does dnscontrol / this provider handle that design choose probably?

das7pad commented:
I am not aware of this particular issue.

Looking at the parsing logic, the value `"foo" "bar" "baz"` should get parsed as expected. That is into a TxtMulti record with `RecordConfig.TxtStrings = ["foo", "bar", "baz"]` via [`RecordConfig.PopulateFromString()->RecordConfig.SetTargetTXTString()->ParseQuotedTxt()`](https://github.com/StackExchange/dnscontrol/blob/de4455942bf38dc6b49e03b56748b00be5c062cf/models/quotes.go#L27-L37).

The serializing logic however is just emitting the first chunk `foo` from the previous example. I noticed that when using `RecordConfig.GetTargetCombined()` to get the full value of a TXT record, the API would respond with a cryptic error for one of the test cases. I cannot test any further at this time as my account is rate-limited :shrug: .  https://github.com/StackExchange/dnscontrol/blob/913d944dbed32bad872c5e548d53fc47bcdb2f4f/providers/hetzner/types.go#L63-L72

I already marked the provider as not capable of handling TxtMulti records. But maybe we can get to the bottom of this validation issue and enable the flag again.
https://github.com/StackExchange/dnscontrol/blob/913d944dbed32bad872c5e548d53fc47bcdb2f4f/providers/hetzner/hetznerProvider.go#L24

tlimoncelli commented on providers/hetzner/hetznerProvider.go:
```diff
@@ -0,0 +1,168 @@
+	for _, m := range del {
+		record := m.Existing.Original.(*Record)
+		corr := &models.Correction{
+			Msg: fmt.Sprintf("%s, HETZNER ID: %s", m.String(), record.Id),
```
Are the HETZNET ID's long?  If that output would look messy, maybe it would be cleaner to either not include it or just include it as `(%s)`

tlimoncelli commented on providers/hetzner/types.go:
```diff
@@ -0,0 +1,88 @@
+		// Cannot use `in.GetTargetCombined()` for TXTs:
+		// Their validation would complain about a missing `;`.
+		// Test case: single_TXT:Create_a_255-byte_TXT
+		// {"error":{"message":"422 Unprocessable Entity: missing: ; ","code":422}}
```
Great comment!  It makes clear to the next dev that looks at the code what is going on in a way that permits them to continue progress.

das7pad commented on providers/hetzner/api.go:
```diff
@@ -0,0 +1,174 @@
+	zones  map[string]Zone
+}
+
+func checkIsLockedSystemRecord(record Record) error {
```
> Why is this called "locked"?

The GUI/web app shows a padlock next to the SOA record.

>  Would checkIsUnsupportedRecord be better?

I do not have a strong opinion on that.

tlimoncelli replied:
Ok, that's fine.

das7pad commented:
Thank you for the review, @tlimoncelli !

All items but the _locked/unsupported_ naming should be addressed by https://github.com/StackExchange/dnscontrol/pull/904/commits/a0f8fd9c8ccf13a2ded97b6694d4f1bcf26d79bd. I could not test the changes yet as my account is still rate-limited. I will try again tomorrow.

tlimoncelli commented:
If you get rate limited, here are some tips:

* Run fewer tests at a time:

The `-start` and `-end` flags control where the tests start and end

Ex: run tests 5 to 10

    go test -v -verbose -provider GANDI -start 5 -end 10

* Slow down the tests.

Add a sleep to the test loop in `integrationTest/integration_test.go` at the start of `makeChanges()`

```
 func makeChanges(t *testing.T, prv providers.DNSServiceProvider, dc *models.DomainConfig, tst *TestCase, desc string, expectChanges bool, origConfig map[string]string) bool {
+       time.Sleep(2 * time.Second)
        domainName := dc.Name
```

* Slow down the tests manually

Sometimes I just press CTRL-Z then run `sleep 10 ; fg` to pause and restart the tests.

das7pad commented:
Taking a closer look at which requests were rate-limited, I noticed that any first request was processed and only further requests were rejected. Apparently their rate-limiting is now set to 1 request per second -- at least for my account.
https://github.com/StackExchange/dnscontrol/pull/904/commits/add8017a890cfd9851b1d0e8b4533f479fb28bfd does some basic rate-limiting aware delaying of requests as soon as we observe a 429. I also added a settings option for enabling it by default `"rate_limited": "true"`. The integration tests pass with and without enabling it by default. In case they lower the rate of 1 req/s, I also added a basic exponential back-off strategy.

I sent them a support request on Friday night about resetting my quota. Maybe this strict rate-limit does not apply to all customers, unless they run the integration test-suite of this tool a couple dozen times :) I also asked them about maintaining this provider.

tlimoncelli commented:
Lookin' good!

golint has some complaints and then I think we're good to go:

```
providers/hetzner/api.go:14:2: const baseUrl should be baseURL
providers/hetzner/api.go:191:6: exported type RequestRateLimiter should have comment or be unexported
providers/hetzner/api.go:213:9: if block ends with a return statement, so drop this else and outdent its block
providers/hetzner/types.go:12:2: struct field ZoneId should be ZoneID
providers/hetzner/types.go:37:6: exported type Record should have comment or be unexported
providers/hetzner/types.go:38:2: struct field Id should be ID
providers/hetzner/types.go:43:2: struct field ZoneId should be ZoneID
providers/hetzner/types.go:46:6: exported type Zone should have comment or be unexported
providers/hetzner/types.go:47:2: struct field Id should be ID
```

das7pad commented:
> Lookin' good!
> 
> golint has some complaints and then I think we're good to go:

Thanks for checking. I used the command [from the docs](https://github.com/StackExchange/dnscontrol/blob/d52c98782c0742222355bc7adce4de323ac23bf5/docs/writing-providers.md#step-11-clean-up), but apparently it's missing the source path.

<details>

```
$ golint
$ echo $?
0
$ golint providers/hetzner/
providers/hetzner/api.go:14:2: const baseUrl should be baseURL
providers/hetzner/api.go:191:6: exported type RequestRateLimiter should have comment or be unexported
providers/hetzner/api.go:213:9: if block ends with a return statement, so drop this else and outdent its block
providers/hetzner/types.go:12:2: struct field ZoneId should be ZoneID
providers/hetzner/types.go:37:6: exported type Record should have comment or be unexported
providers/hetzner/types.go:38:2: struct field Id should be ID
providers/hetzner/types.go:43:2: struct field ZoneId should be ZoneID
providers/hetzner/types.go:46:6: exported type Zone should have comment or be unexported
providers/hetzner/types.go:47:2: struct field Id should be ID
```
</details>

golint complaints are addressed in https://github.com/StackExchange/dnscontrol/pull/904/commits/59a5e31ad615b272ae069f8b6c938dc8d654750e . Final touches to the rate-limiting are in https://github.com/StackExchange/dnscontrol/pull/904/commits/a2882dd8c53cbda3b0f5aadaf988f1ac79a66119 and https://github.com/StackExchange/dnscontrol/pull/904/commits/db9de4cb97991da1210ea9f02455c8ad246b1ace.
The integration tests pass. Manual testing via a locally built docker-image and a fairly diverse 'real-world' zone were successful as well. The output of `get-zones ... HETZNER all` is looking good too.

tlimoncelli commented:
Thank you for contributing this provider!

das7pad commented:
Thank you for the guidance!

Hetzner replied to me on Wednesday. They adjusted the rate-limiting for my account. I do not think this helps much as we need a solution that works for all users. I will not add multiple tiers of rate-limits to this provider. They are happy about this integration, but do not seem to want to maintain it (they avoided to answer this question).

tlimoncelli commented:
Yeah, vendors rarely commit to maintaining something in an open source project unless there is a specific reason.

If you want to implement a back-off algorithm it isn't too difficult.  Basically the 429 error (or whatever the use to indicate you've gone too fast) indicates you should sleep then resend the exact same request.

There is a good example in `providers/desec/protocol.go` as well as `providers/namecheap/namecheapProvider.go` and `providers/route53/route53Provider.go`.

If you want to be fancy, the 429 response comes with a `Retry-After: 3600` header that tells you how long to sleep, but most dnscontrol providers just sleep half a second.

das7pad commented:
> If you want to implement a back-off algorithm it isn't too difficult. Basically the 429 error (or whatever the use to indicate you've gone too fast) indicates you should sleep then resend the exact same request.

I already have [a back-off strategy in-place](https://github.com/StackExchange/dnscontrol/blob/61bf50d4d7bc7b9c30627d7d8c77d65a300ad4c9/providers/hetzner/api.go#L213-L227). But it's not very smart and it falls short for their rate-limiting which is based on strict quotas in multiple _buckets_ per hour, per minute, per second. Presumably a sliding window, so after an hour/minute/second everything resets, but a large upload may still run with minutes of delays in between requests.
 
> There is a good example in `providers/desec/protocol.go` as well as `providers/namecheap/namecheapProvider.go` and `providers/route53/route53Provider.go`.
> 
> If you want to be fancy, the 429 response comes with a `Retry-After: 3600` header that tells you how long to sleep, but most dnscontrol providers just sleep half a second.

Thanks for the pointers, I took a look at the underlying tool they use for rate-limiting, their api proxy `kong`. It broadcasts the exact limits for each quota (hour, minute and second) https://docs.konghq.com/hub/kong-inc/rate-limiting/#headers-sent-to-the-client. With these limits we can plan the request delays to a certain extent and stay _under the radar_.
https://github.com/das7pad/dnscontrol/commit/d58d24ffdf264247703616ee780cec45a6795a3e implements and documents all of it. I will open a PR after #925 landed, as these would conflict (git-wise) at this time.
